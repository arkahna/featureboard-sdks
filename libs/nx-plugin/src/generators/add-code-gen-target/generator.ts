import type { TargetConfiguration, Tree } from '@nx/devkit'
import {
    readProjectConfiguration,
    updateProjectConfiguration,
} from '@nx/devkit'
import type { CodeGenExecutorSchema } from '../../executors/code-gen/schema'
import type { AddCodeGenTargetGeneratorSchema } from './schema'

export async function addCodeGenTargetGenerator(
    tree: Tree,
    options: AddCodeGenTargetGeneratorSchema,
) {
    const project = readProjectConfiguration(tree, options.projectName)
    project.targets ??= {}

    project.targets[options.targetName ?? 'code-gen'] = {
        ...getCodeGenExecutorConfiguration(options),
    }

    updateProjectConfiguration(tree, options.projectName, project)
}

/**
 * Returns a TargetConfiguration for the @featureboard/nx-plugin:code-gen executor
 */
export function getCodeGenExecutorConfiguration(
    options: CodeGenExecutorSchema,
): CodeGenExecutorConfiguration {
    const codeGenOptions: CodeGenExecutorConfiguration['options'] = {
        template: options.template,
        featureBoardProductName: options.featureBoardProductName,
        subFolder: options.subFolder,
    }
    if (options.dryRun) {
        codeGenOptions.dryRun = true
    }
    return {
        executor: '@featureboard/nx-plugin:code-gen',
        outputs: [`{projectRoot}/${options.subFolder}`.replace('/./', '')],
        options: codeGenOptions,
        dependsOn: ['build'],
    }
}

/**
 * Configuration options relevant for the executor
 */
export type CodeGenExecutorConfiguration = TargetConfiguration & {
    executor: '@featureboard/nx-plugin:code-gen'
    options: CodeGenExecutorSchema & { dryRun?: boolean }
}

export default addCodeGenTargetGenerator
