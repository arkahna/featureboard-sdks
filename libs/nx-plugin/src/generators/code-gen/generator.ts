import {
    getValidToken,
    readCurrentOrganization,
} from '@featureboard/api-authentication'
import type { FeatureBoardAuth } from '@featureboard/code-generator'
import { codeGenerator } from '@featureboard/code-generator'
import type { Tree } from '@nx/devkit'
import { joinPathFragments, readProjectConfiguration } from '@nx/devkit'
import { API_ENDPOINT, CLIENT_ID } from '../../shared/config'
import { isDryRun } from '../../shared/is-dry-run'
import { isVerbose } from '../../shared/is-verbose'
import type { CodeGenGeneratorSchema } from './schema'

export async function codeGenGenerator(
    tree: Tree,
    { projectName, ...options }: CodeGenGeneratorSchema,
): Promise<void> {
    const project = readProjectConfiguration(tree, projectName)
    const dryRun = isDryRun()
    const verbose = isVerbose()

    let auth: FeatureBoardAuth
    if (options.featureBoardApiKey) {
        auth = {
            featureBoardApiKey: options.featureBoardApiKey,
        }
    } else {
        const featureBoardBearerToken = await getValidToken(CLIENT_ID)
        const organizationId = await readCurrentOrganization(verbose)
        if (!featureBoardBearerToken || !organizationId) {
            return
        }

        auth = {
            featureBoardBearerToken,
            organizationId,
        }
    }

    await codeGenerator({
        tree: tree,
        relativeFilePath: joinPathFragments(project.root, options.subFolder),
        interactive: !dryRun,
        auth,
        apiEndpoint: API_ENDPOINT,
        ...options,
    })
}

export default codeGenGenerator
