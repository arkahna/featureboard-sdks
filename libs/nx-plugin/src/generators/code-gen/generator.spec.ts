import type { Tree } from '@nx/devkit'
import {
    addProjectConfiguration,
    joinPathFragments,
    readProjectConfiguration,
} from '@nx/devkit'
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing'
import * as fs from 'fs/promises'
import { HttpResponse, http } from 'msw'
import type { SetupServer } from 'msw/node'
import { setupServer } from 'msw/node'
import * as path from 'path'
import { codeGenGenerator } from './generator'
import type { CodeGenGeneratorSchema } from './schema'

describe('code-generator', () => {
    let tree: Tree
    let options: CodeGenGeneratorSchema
    let server: SetupServer

    beforeAll(() => {
        server = setupServer(
            http.get('https://api.featureboard.app/projects', async () => {
                const file = await fs.readFile(
                    path.join(__dirname, '../../../test-data/projects.json'),
                    {
                        encoding: 'utf8',
                    },
                )
                return HttpResponse.json(JSON.parse(file))
            }),
            http.get(
                'https://api.featureboard.app/projects/ivu5ntxbxeeorm248jl25bs6/features',
                async () => {
                    const file = await fs.readFile(
                        path.join(
                            __dirname,
                            '../../../test-data/features-ivu5ntxbxeeorm248jl25bs6.json',
                        ),
                        {
                            encoding: 'utf8',
                        },
                    )
                    return HttpResponse.json(JSON.parse(file))
                },
            ),
        )
        server.listen({ onUnhandledRequest: 'error' })
    })

    beforeEach(async () => {
        tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' })

        options = {
            projectName: 'my-app',
            subFolder: './',
            template: 'dotnet-api',
            featureBoardProductName: 'SaaSy Icons',
            featureBoardApiKey: 'This is totally a key',
        }

        const root = 'apps/my-app'
        addProjectConfiguration(tree, options.projectName, {
            root: `apps/${options.projectName}`,
            projectType: 'application',
            targets: { 'my-target': { executor: 'nx:noop' } },
        })
        tree.write(
            joinPathFragments(root, 'Project.FeatureBoard.MyApp.csproj'),
            '',
        )
    })

    describe('template: dotnet-api', () => {
        beforeEach(async () => {
            options.template = 'dotnet-api'
        })

        it('should run successfully', async () => {
            let project = readProjectConfiguration(tree, options.projectName)
            await codeGenGenerator(tree, options)

            project = readProjectConfiguration(tree, options.projectName)
            expect(project).toBeDefined()
        })

        it('should produce the expected features files', async () => {
            let project = readProjectConfiguration(tree, options.projectName)
            await codeGenGenerator(tree, options)

            project = readProjectConfiguration(tree, options.projectName)
            const files = tree
                .listChanges()
                .filter(
                    (x) =>
                        x.path.match(/^apps\/my-app\/.*$/) &&
                        !x.path.endsWith('.csproj') &&
                        !x.path.endsWith('/project.json'),
                )

            expect(files).toHaveLength(1)
            files.forEach((x) => {
                expect(x.content?.toString('utf-8')).toMatchSnapshot(x.path)
            })
        })
    })

    describe('template: typescript', () => {
        beforeEach(async () => {
            options.template = 'typescript'
        })

        it('should run successfully', async () => {
            let project = readProjectConfiguration(tree, options.projectName)
            await codeGenGenerator(tree, options)

            project = readProjectConfiguration(tree, options.projectName)
            expect(project).toBeDefined()
        })

        it('should produce the expected features files', async () => {
            let project = readProjectConfiguration(tree, options.projectName)
            await codeGenGenerator(tree, options)

            project = readProjectConfiguration(tree, options.projectName)
            const files = tree
                .listChanges()
                .filter(
                    (x) =>
                        x.path.match(/^apps\/my-app\/.*$/) &&
                        !x.path.endsWith('.csproj') &&
                        !x.path.endsWith('/project.json'),
                )

            expect(files).toHaveLength(1)
            files.forEach((x) => {
                expect(x.content?.toString('utf-8')).toMatchSnapshot(x.path)
            })
        })
    })

    afterEach(() => {
        server.resetHandlers()
    })

    afterAll(() => {
        server.close()
    })
})
