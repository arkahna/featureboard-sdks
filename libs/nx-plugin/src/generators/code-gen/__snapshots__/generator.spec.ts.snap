// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`code-gen generator > template: dotnet-api > should produce the expected features files > apps/my-app/Features.cs 1`] = `
"using FeatureBoard.DotnetSdk.Attributes;
using FeatureBoard.DotnetSdk.Models;
using System.Runtime.Serialization;

namespace Project.FeatureBoard.MyApp;

// Features
public class Features : IFeatures
{
  [FeatureKeyName(\\"requires-attribution\\")]
  public bool RequiresAttribution { get; set; }
  [FeatureKeyName(\\"allow-edits\\")]
  public bool AllowEdits { get; set; }
  [FeatureKeyName(\\"number-custom-collections\\")]
  public decimal NumberCustomCollections { get; set; }
  [FeatureKeyName(\\"team-management\\")]
  public bool TeamManagement { get; set; }
  [FeatureKeyName(\\"custom-collections\\")]
  public CustomCollections CustomCollections { get; set; }
  [FeatureKeyName(\\"view-subscribers\\")]
  public bool ViewSubscribers { get; set; }
}


// Options Enums

//Custom collections
public enum CustomCollections
{
  [EnumMember(Value = \\"Disabled\\")]
  Disabled,
  [EnumMember(Value = \\"Enabled\\")]
  Enabled,
  [EnumMember(Value = \\"Shared\\")]
  Shared,
}

// Type Enums
public enum BooleanFeature
{
  RequiresAttribution,
  AllowEdits,
  TeamManagement,
  ViewSubscribers,
}
public enum NumberFeature
{
  NumberCustomCollections,
}
public enum StringFeature
{
}

// Feature Gate attribute
[AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
public class FeatureFilterAttribute : FeatureFilterAttributeBase
{
  protected override string Feature { get; }
  protected override object DefaultValue { get; }
  protected override object? IsEqualTo { get; }
  private readonly Dictionary<string, string> _featureKeyLookup = new()
  {
    { \\"RequiresAttribution\\", \\"requires-attribution\\" },
    { \\"AllowEdits\\", \\"allow-edits\\" },
    { \\"NumberCustomCollections\\", \\"number-custom-collections\\" },
    { \\"TeamManagement\\", \\"team-management\\" },
    { \\"ViewSubscribers\\", \\"view-subscribers\\" },
  };

  public FeatureFilterAttribute(BooleanFeature feature, bool defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(BooleanFeature feature, bool isEqualTo, bool defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(StringFeature feature, string isEqualTo, string defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(NumberFeature feature, decimal isEqualTo, decimal defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(CustomCollections isEqualTo, CustomCollections defaultValue)
  {
    Feature = \\"custom-collections\\";
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

}
"
`;

exports[`code-gen generator > template: typescript > should produce the expected features files > apps/my-app/features.ts 1`] = `
"import '@featureboard/js-sdk'

export interface SaaSyIconsFeatures {

    /**
    * Requires attribution
    * @description 
    */
    'requires-attribution': boolean

    /**
    * Allow edits
    * @description Allows the user to edit colour, backgrounds
    */
    'allow-edits': boolean

    /**
    * Number custom collections
    * @description 
    */
    'number-custom-collections': number

    /**
    * Team management
    * @description 
    */
    'team-management': boolean

    /**
    * Custom collections
    * @description 
    */
    'custom-collections': 'Disabled' | 'Enabled' | 'Shared'

    /**
    * View subscribers
    * @description 
    */
    'view-subscribers': boolean
}

declare module '@featureboard/js-sdk' {
    // eslint-disable-next-line @typescript-eslint/no-empty-interface
    interface Features extends SaaSyIconsFeatures {}
}

export type SaaSyIconsFeature = keyof SaaSyIconsFeatures"
`;
